doc/man-source/irpty.1
.TH IRPTY "1" "Last change: Aug 2015" "irpty @version@" "User Commands"
.SH NAME
.P
\fBirpty\fR - Pseudo tty driver.
.SH SYNOPSIS
.B irpty
[\fIoptions\fR] \fIconfig_file -- program \fR[\fIargs \fR...]
.SH DESCRIPTION
.P
irpty connects to lircd to receive infrared codes and converts them to
key strokes. E.g. type irpty ~/.config/lircrc \-\- workbone to control the
CD-player program workbone. Of course you will have to create an
appropriate config file for this purpose first. The config string will
be passed to the desired application. Note that you can use escape
sequences to specify non printable characters. Have a look at
the .lircrc file format description for details.
.SH OPTIONS
.TP
\fB\-h\fR \fB\-\-help\fR
display usage summary
.TP
\fB\-V\fR \fB\-\-version\fR
display version
.TP
\fB\-e\fR \fB\-\-no\-echo\fR
disable echo
.TP
\fB\-i\fR \fB\-\-ignore\-eof\fR
ignore EOF
.TP
\fB\-n\fR \fB\-\-non\-interactive\fR
force non\-interactive mode
.TP
\fB\-v\fR \fB\-\-verbose\fR
verbose mode
.SH "SEE ALSO"
.P
The documentation for \fBlirc\fR
is maintained as html pages. They are located under html/ in the
documentation directory.


NEWS
0.9.4

  * The select() calls in lircd, lircrcd and library routines have
    been replaced with poll() to avoid crashes when file descriptors
    becomes > 1024 (#109).
  * The documentation is heavily revised. Old obsolete text is
    removed, the manual is reorganized, some info is moved to
    manpages and a lot of minor updates and bugfixes.
  * The remotes/ directory is not used any more, removed.
  * Set supplementary groups using effective-user (RH bz 1253907).
  * Work around kernel bug in /dev/lirc0 output (#119).

0.9.3pre2 1/08/15
  * irtoy: New support for debug LEDs. New documentation
    file, support in lirc-setup and driver table doc added.
  * The library routines for parsing lircd.conf now warns for
    duplicate keys and values (#107).
  * lirc-lsremotes now displays logs after parsing files, making
    more parsing warnings visible.
  * lirc-config-tool is fixed to comply with guidelines (#105).
  * Bug in irexec when using modes fixed (#102).
  * Options parsing (notably -U) and just listing first remote
    bugs in irsimsend have been fixed.
  * A new kernel driver irpipe has been added. This allows feeding data
    from userspace to lircd, which can use the /dev/irpipeX device
    as a clean substitute for /dev/lircX. A companion CLI tool
    irpipe(1) is also added.
0
0.9.3pre1 30/06/15

  * The irexec semantics have been changed so that the invoked
    programs run disconnected from the main irexec process.
    See irexec(1) (#92).
  * mode2 has been updated to handle also LIRCCODE devices (#94)
    and the --driver-options switch
  * xmode2 has been updated to handle udp driver, the new
    --driver-options switch + some general overhaul.
  * irrecord has got some love:
     + It handles multiple definitions of same symbol OK (#78)
     + Warns about single-key config files (#77)
     + Checks for ambient light disturbances (#76)
     + New option allows updating of existing template file.
     + Template handling bug fixed
     + -l --list-namespace option parsing bug fixed (#101).
  * irrecord has been split into a library and a cli interface
    (#36). The library interface is unstable.
  * lircmd: Several option parsing bugs fixed, one of which
    blocked the systemd service (#104, RH bz 1234155).
  * lirc-setup: fixed name clash on recent python versions.
  * The udp driver is patched to allow adjustable clockrates,
    see new udp_driver.html driver doc.
  * The mplay driver has been updated with patches which has
    been available.
  * Overall code formatting is improved with more consistent
    layout and updated git hook. See CONTRIBUTE.md
  * The compiler has been switched to c++, and filenames are
    updated to *.cpp (#79)
  * New build dependency: xsltproc (usually coming with libxml).
  * Unlinked and hard to find driver docs and manpages have
    been linked.
  * Build: now uses parallell make by default.
  * Docs: The Configuration Guide has been updated.

0.9.2 12/09/14
  * Documentation fixes

0.9.2pre3  12/04/14
  * Add fix for #70, by mistake omitted in pre2.

0.9.2pre2  11/30/14
  * Fixes for #67 - #75 besides #71 and #74.

0.9.2pre1  11/21/14
  * Removed functionality:
    + The smode2 program has been retired from the build, although it still
      is available in contrib/.
    + The setup.sh script is removed (build always builds all drivers).
  * Drivers are loaded dynamically. Many programs including lircd now
    needs to know where the drivers are located to be able to use
    them. There are new options, environment variables and hard-coded
    defaults. See the manpages for more.
  * A new tool lirc-lsplugins is added to inspect dynamic driver modules.
  * New, unstable configuration script lirc-setup. The script guides user
    when selecting both capture device and remote config. It sports auto-
    detection of most devices, searching and downloading from the remotes
    database and a considerable amount of bugs.
  * The remotes/ directory has become obsolete, only used for testing.
    New configuration tools uses the remotes database at
    http://sf.net/p/lirc-remotes as the only source (#53). A new tool
    irdb-get(1) supports searching and downloading from remotes database.
  * Building on MacOSX has been refreshed.
  * New devices and features:
    + Support for the Dangerous Prototypes irtoy device is added (#20).
    + Support for for sonyir on MacOSX is added. This is a single hiddev
      device; with the framework in place hopefully more devices will be
      added over time.
    + Support for the DIY assembly kit yard2 usb ir device is added. This is
      an experimental driver which depends on a server which currently is not
      packaged in any major distro; a Debian/Ubuntu PPA is available.
    + Support for the Roku remote protocol added (#8).
    + Support for long keypresses has been added. See the "Configuring LIRC"
      chapter in the manual, search for "ignore_first_event" (#50).
  * lircd has been verified to run as regular user as long as it can open
    devices and such. A new option --effective-user makes lircd drop it's
    privileges after opening devices if started as root. A new udev rule
    is shipped to make /dev/lirc* accessible for users in group 'lirc" (#56).
  * Logging runtime option is renamed: debug -> loglevel. Levels can be given
    symbolically e. g., 'info' or 'error'. Old option name is still supported.
  * Configuration changes
    + The default location for the user lircrc-file has been changed from
      ~/.lircrc to ~/.config/lircrc. The old location is still used as fallback.
      (ticket #52).
    + The lircd.conf configuration has been updated. lircd will now besides
      the lircd.conf file also read all *.conf files in the
      /etc/lirc/lircd.conf.d directory. The already existing include directive
      has been modified to accept wildcards and also been documented.
    + lircd.conf has a new 'driver' attribute for configs requiring a specific

tools/irpty.cpp
/****************************************************************************
** irpty.c *****************************************************************
****************************************************************************
*
* irpty  - pseudo tty driver
*          Connects to lircd via socket to receive infra-red codes
*          and converts them to key strokes
*
* Copyright (C) 1996,97 Ralph Metzler <rjkm@thp.uni-koeln.de>
* Copyright (C) 1998 Christoph Bartelmus <lirc@bartelmus.de>
*
*/

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <grp.h>
#include <poll.h>
#include <signal.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <termios.h>
#include <unistd.h>

#ifdef HAVE_FORKPTY
#if defined __APPLE__ || defined __NetBSD__
#include <util.h>
#else
#if defined __FreeBSD__
#include <libutil.h>
#else
#include <pty.h>
#endif
#endif
#endif

#include "lirc_client.h"

#define BUFFSIZE 512

const char* progname = "irpty";

struct lirc_config* lconfig;

static int lsock, sigcaught;

void die(const char* fmt, ...)
{
	va_list args;

	va_start(args, fmt);
	vfprintf(stderr, fmt, args);
	lirc_freeconfig(lconfig);
	lirc_deinit();
	exit(1);
}

static void sig_term(int sig)
{
	sigcaught = 1;
}

static void copy_loop(int ptym, int ignoreeof)
{
	pid_t child;
	int nread;
	char buf[BUFFSIZE];
	struct sigaction act;
	struct pollfd pfd[2] = {
		{lsock, POLLIN, 0},
		{STDIN_FILENO, POLLIN, 0}
                // fd, events, revents
	};;

	if ((child = fork()) < 0) {
		die("fork error");
	} else if (!child) {

		while (1) {
			poll(pfd, 2, 0);

			if (pfd[1].revents & POLLIN) { /* STDIN_FILENO */
				if ((nread = read(STDIN_FILENO, buf, BUFFSIZE)) < 0)
					die("read error from stdin");
				else if (!nread)
					break;
				if (write(ptym, buf, nread) != nread)
					die("writen error to master pty");
			}
			if (pfd[0].revents & POLLIN) {  /*lsock */
				char* ir;
				char* irchars;
				int ret;

				while ((ret = lirc_nextcode(&ir)) == 0) {
					if (ir == NULL)
						break;
					while ((ret = lirc_code2char(lconfig, ir, &irchars)) == 0 && irchars != NULL)
						if (write(ptym, irchars, strlen(irchars)) != (int) strlen(irchars))
							die("writen error to master pty");
					free(ir);
					if (ret == -1)
						break;
				}
				if (ret == -1)
					break;
			}
		}
		if (!ignoreeof)
			kill(getppid(), SIGTERM);
		lirc_freeconfig(lconfig);
		lirc_deinit();
		_exit(0);
	}

	act.sa_handler = sig_term;
	sigemptyset(&act.sa_mask);
	act.sa_flags = 0;       /* we need EINTR */
	sigaction(SIGTERM, &act, NULL);

	while (1) {
		if ((nread = read(ptym, buf, BUFFSIZE)) <= 0)
			break;
		if (write(STDOUT_FILENO, buf, nread) != nread)
			die("writen error to stdout");
	}
	if (!sigcaught)
		kill(child, SIGTERM);
	lirc_freeconfig(lconfig);
	lirc_deinit();
}

static struct termios save_termios;
static int ttysavefd = -1;
static enum {
	RESET, RAW, CBREAK
} ttystate = RESET;

int tty_raw(int fd)
{
	struct termios buf;

	if (tcgetattr(fd, &save_termios) < 0)
		return -1;

	buf = save_termios;
	buf.c_lflag &= ~(ECHO | ICANON | IEXTEN | ISIG);
	buf.c_iflag &= ~(BRKINT | ICRNL | INPCK | ISTRIP | IXON);
	buf.c_cflag &= ~(CSIZE | PARENB);
	buf.c_cflag |= CS8;
	buf.c_oflag &= ~(OPOST);
	buf.c_cc[VMIN] = 1;
	buf.c_cc[VTIME] = 0;

	if (tcsetattr(fd, TCSAFLUSH, &buf) < 0)
		return -1;
	ttystate = RAW;
	ttysavefd = fd;
	return 0;
}

int tty_reset(int fd)
{
	if (ttystate != CBREAK && ttystate != RAW)
		return 0;
	if (tcsetattr(fd, TCSAFLUSH, &save_termios) < 0)
		return -1;
	ttystate = RESET;
	return 0;
}

void tty_atexit(void)
{
	if (ttysavefd >= 0)
		tty_reset(ttysavefd);
}

#ifdef HAVE_FORKPTY

#define pty_fork(ptrfdm, slave_name, slave_termios, slave_winsize) \
	forkpty(ptrfdm, slave_name, slave_termios, slave_winsize)

#else
/* Open the next free pty */

int pty_open(char* pty_name)
{
	char* ptr1;
	char* ptr2;
	int fd;

	strcpy(pty_name, "/dev/ptyp0");
	for (ptr1 = "pqrstuvwxyzabcde"; *ptr1; ptr1++) {
		pty_name[8] = *ptr1;
		for (ptr2 = "0123456789abcdef"; *ptr2; ptr2++) {
			pty_name[9] = *ptr2;

			if ((fd = open(pty_name, O_RDWR)) >= 0) {
				pty_name[5] = 't';
				return fd;
			} else if (errno == ENOENT) {
				return -1;
			}
		}
	}
	return -1;
}

int tty_open(int fdm, char* tty_name)
{
	struct group* grptr;
	int gid, fds;

	if ((grptr = getgrnam("tty")) != NULL)
		gid = grptr->gr_gid;
	else
		gid = -1;

	/*
	 * chown(tty_name, getuid(), gid);
	 * chmod(tty_name, S_IRUSR | S_IWUSR | S_IWGRP);
	 */

	if ((fds = open(tty_name, O_RDWR)) < 0) {
		close(fdm);
		return -1;
	}
	return fds;
}

pid_t pty_fork(int* ptrfdm, char* slave_name, struct termios* slave_termios, struct winsize* slave_winsize)
{
	int fdm, fds;
	pid_t pid;
	char pts_name[20];

	if ((fdm = pty_open(pts_name)) < 0)
		die("can't open pty %s\n", pts_name);
	if (slave_name)
		strcpy(slave_name, pts_name);

	if ((pid = fork()) < 0)
		die("fork error\n");
	if (!pid) {
		if (setsid() < 0)
			die("setsid error");
		if ((fds = tty_open(fdm, pts_name)) < 0)
			die("can't open slave pty %s\n", pts_name);
		close(fdm);

		if (slave_termios)
			if (tcsetattr(fds, TCSANOW, slave_termios) < 0)
				die("tcsetattr error on slave pty");
		if (slave_winsize)
			if (ioctl(fds, TIOCSWINSZ, slave_winsize) < 0)
				die("TIOCSWINSZ error on slave pty");
		if (dup2(fds, STDIN_FILENO) != STDIN_FILENO)
			die("dup2 error to stdin");
		if (dup2(fds, STDOUT_FILENO) != STDOUT_FILENO)
			die("dup2 error to stdout");
		if (dup2(fds, STDERR_FILENO) != STDERR_FILENO)
			die("dup2 error to stderr");
		if (fds > STDERR_FILENO)
			close(fds);
		return 0;
	}
	*ptrfdm = fdm;
	return pid;
}
#endif

static void set_noecho(int fd)
{
	struct termios stermios;

	if (tcgetattr(fd, &stermios) < 0)
		die("tcgetattr error");
	stermios.c_lflag &= ~(ECHO | ECHOE | ECHOK | ECHONL);
	stermios.c_oflag &= ~(ONLCR);
	if (tcsetattr(fd, TCSANOW, &stermios) < 0)
		die("tcsetattr error");
}

static struct option long_options[] = {
	{ "help",	     no_argument,	NULL, 'h' },
	{ "version",	     no_argument,	NULL, 'V' },
	{ "socket",	     required_argument, NULL, 's' },
	{ "no-echo",	     no_argument,	NULL, 'e' },
	{ "ignore-eof",	     no_argument,	NULL, 'i' },
	{ "non-interactive", no_argument,	NULL, 'n' },
	{ "verbose",	     no_argument,	NULL, 'v' },
	{ 0,		     0,			0,    0	  }
};

int main(int argc, char* argv[])
{
	int fdm, c, ignoreeof, interactive, noecho, verbose;
	pid_t pid;
	char* config;
	char slave_name[20];
	struct termios orig_termios;
	struct winsize size;
	int flags;

	interactive = isatty(STDIN_FILENO);
	ignoreeof = 0;
	noecho = 0;
	verbose = 0;
	config = NULL;

	while ((c = getopt_long(argc, argv, "hVs:einv", long_options, NULL))
	       != EOF) {
		switch (c) {
		case 'h':
			printf("Usage: %s [options] config_file -- program [args ...]\n", argv[0]);
			printf("\t -h --help \t\tdisplay usage summary\n");
			printf("\t -V --version \t\tdisplay version\n");
			printf("\t -e --no-echo \t\tdisable echo\n");
			printf("\t -i --ignore-eof \tignore EOF\n");
			printf("\t -n --non-interactive \tforce non-interactive mode\n");
			printf("\t -v --verbose \t\tverbose mode\n");
			return EXIT_SUCCESS;
		case 'V':
			printf("%s %s\n", progname, VERSION);
			return EXIT_SUCCESS;

		case 'e':
			noecho = 1;
			break;

		case 'i':
			ignoreeof = 1;
			break;

		case 'n':
			interactive = 0;
			break;

		case 'v':
			verbose = 1;
			break;

		case '?':
			die("unrecognized option: -%c\n", optopt);
		}
	}
	if (optind + 1 >= argc)
		die("usage: irpty [ -s server -einv ] cfg program [ arg ... ]\n");

	config = argv[optind++];

	if ((lsock = lirc_init("irpty", 1)) == -1)
		exit(EXIT_FAILURE);
	flags = fcntl(lsock, F_GETFL, 0);
	if (flags != -1)
		fcntl(lsock, F_SETFL, flags | FASYNC | O_NONBLOCK);

	if (lirc_readconfig(config, &lconfig, NULL) != 0)
		exit(EXIT_FAILURE);

	if (interactive) {
		if (tcgetattr(STDIN_FILENO, &orig_termios) < 0)
			die("tcgetattr error on stdin\n");
		if (ioctl(STDIN_FILENO, TIOCGWINSZ, (char*)&size) < 0)
			die("TIOCGWINSZ error\n");
		pid = pty_fork(&fdm, slave_name, &orig_termios, &size);
	} else {
		pid = pty_fork(&fdm, slave_name, NULL, NULL);
	}

	if (pid < 0) {
		die("fork error\n");
	} else if (!pid) {                              /* child */
		if (noecho)
			set_noecho(STDIN_FILENO);       /* stdin is slave pty */
		if (execvp(argv[optind], &argv[optind]) < 0)
			die("can't execute: %s\n", argv[optind]);
	}
	if (verbose) {
		fprintf(stderr, "slave name = %s\n", slave_name);
		if (config)
			fprintf(stderr, "config file = %s\n", config);
	}
	if (interactive) {
		if (tty_raw(STDIN_FILENO) < 0)  /* user's tty to raw mode */
			die("tty_raw error");
		if (atexit(tty_atexit) < 0)     /* reset user's tty on exit */
			die("atexit error");
	}
	copy_loop(fdm, ignoreeof);

	exit(0);
}
